File: legacy/examples/cpp/core/sample_texture.cpp

#include <CEngine.h>

int main()
{
    auto _app = std::make_unique<engine::CApplication>();

    auto _textureDataJpg = engine::CTextureManager::GetCachedTextureData( "img_container" );
    auto _textureDataPng = engine::CTextureManager::GetCachedTextureData( "img_smiley" );

    if ( _textureDataJpg )
    {
        std::cout << "texture-data-jpg:" << std::endl;
        std::cout << engine::toString( *_textureDataJpg ) << std::endl;
    }

    if ( _textureDataPng )
    {
        std::cout << "texture-data-png:" << std::endl;
        std::cout << engine::toString( *_textureDataPng ) << std::endl;
    }

    auto _textureContainer = engine::CTextureManager::GetCachedTexture( "img_container" );
    ENGINE_ASSERT( _textureContainer, "Could not retrieve valid texture for the sample - container" );

    auto _textureSmiley = engine::CTextureManager::GetCachedTexture( "img_smiley" );
    ENGINE_ASSERT( _textureContainer, "Could not retrieve valid texture for the sample - smiley" );

    float _bufferData[] = {
    /*|      pos     |     color      |  texture  */
        -0.5f, -0.5f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f,
         0.5f, -0.5f, 0.0f, 1.0f, 0.0f, 1.0f, 0.0f,
         0.5f,  0.5f, 0.0f, 0.0f, 1.0f, 1.0f, 1.0f,
        -0.5f,  0.5f, 1.0f, 1.0f, 1.0f, 0.0f, 1.0f
    };

    engine::uint32 _indices[] = {
        0, 1, 2, // first triangle
        0, 2, 3  // second triangle
    };

    engine::CVertexBufferLayout _layout = { { "pos", engine::eElementType::Float2, false },
                                            { "col", engine::eElementType::Float3, false },
                                            { "uv",  engine::eElementType::Float2, false } };

    auto _vbuffer = std::make_unique<engine::CVertexBuffer>( _layout, engine::eBufferUsage::STATIC,
                                                             sizeof( _bufferData ), _bufferData );
    auto _ibuffer = std::make_unique<engine::CIndexBuffer>( engine::eBufferUsage::STATIC, 6, _indices );

    auto _varray = std::make_unique<engine::CVertexArray>();
    _varray->addVertexBuffer( std::move( _vbuffer ) );
    _varray->setIndexBuffer( std::move( _ibuffer ) );

    auto _shader = engine::CShaderManager::GetCachedShader( "basic2d_textures" );

    while( _app->active() )
    {
        tinyutils::Clock::Tick();
        _app->begin();

        if ( engine::CInputManager::IsKeyDown( engine::Keys::KEY_ESCAPE ) )
            break;

        _shader->bind();
        _varray->bind();
        // _textureContainer->bind();
        _textureSmiley->bind();

        glDrawElements( GL_TRIANGLES, _varray->indexBuffer()->count(), GL_UNSIGNED_INT, 0 );

        _textureSmiley->unbind();
        // _textureContainer->unbind();
        _varray->unbind();
        _shader->unbind();

        _app->end();
        tinyutils::Clock::Tock();
    }

    return 0;
}

File: legacy/examples/python/core/sample_texture.py
#!/usr/bin/env python

from tinyrenderer.core import engine
import numpy as np
from OpenGL.GL import *

if __name__ == '__main__' :
    app = engine.Application()

    textureDataJpg = engine.TextureManager.GetCachedTextureData( 'img_container' )
    textureDataPng = engine.TextureManager.GetCachedTextureData( 'img_smiley' )

    if ( textureDataJpg ) :
        print( 'texture-data-jpg: \n\r{}'.format( textureDataJpg ) )

    if ( textureDataPng ) :
        print( 'texture-data-png: \n\r{}'.format( textureDataPng ) )

    textureContainer = engine.TextureManager.GetCachedTexture( 'img_container' )
    assert textureContainer, 'Could not retrieve a valid texture for the sample: container'

    textureSmiley = engine.TextureManager.GetCachedTexture( 'img_smiley' )
    assert textureSmiley, 'Could not retrieve a valid texture for the sample: smiley'

    bufferData = np.array( [
     #|    pos    |     color    |  texture  |#
        -0.5, -0.5, 1.0, 0.0, 0.0, 0.0, 0.0,
         0.5, -0.5, 0.0, 1.0, 0.0, 1.0, 0.0,
         0.5,  0.5, 0.0, 0.0, 1.0, 1.0, 1.0,
        -0.5,  0.5, 1.0, 1.0, 1.0, 0.0, 1.0
    ], dtype = np.float32 )

    indices = np.array( [
        0, 1, 2, # first triangle
        0, 2, 3  # second triangle
    ], dtype = np.uint32 )

    layout = engine.VertexBufferLayout( [ [ 'pos', engine.ElementType.Float2, False ],
                                      [ 'col', engine.ElementType.Float3, False ],
                                      [ 'uv',  engine.ElementType.Float2, False ] ] )

    vbuffer = engine.VertexBuffer( layout,
                               engine.BufferUsage.STATIC,
                               bufferData.size * bufferData.itemsize,
                               bufferData )

    ibuffer = engine.IndexBuffer( engine.BufferUsage.STATIC, 6, indices )

    varray = engine.VertexArray()
    varray.addVertexBuffer( vbuffer )
    varray.setIndexBuffer( ibuffer )

    shader = engine.ShaderManager.GetCachedShader( 'basic2d_textures' )

    while app.active() :
        app.begin()

        if ( engine.InputManager.IsKeyDown( engine.Keys.KEY_ESCAPE ) ) :
            break

        shader.bind()
        varray.bind()
        ## textureContainer.bind()
        textureSmiley.bind()

        if ( varray.indexBuffer() ) :
            glDrawElements( GL_TRIANGLES, varray.indexBuffer().count(), GL_UNSIGNED_INT, None )
        else :
            glDrawArrays( GL_TRIANGLES, 0, 6 )

        textureSmiley.unbind()
        ## textureContainer.unbind()
        varray.unbind()
        shader.unbind()
        app.end()
